src/GHC/RTS/Events/Incremental.hs

```diff
@@ -0,0 +1,202 @@
+  -- continue.
+  | Produce !a (Decoder a)
+  -- ^ The decoder has returned a decoded value and the next decoder state to
+  -- continue.
+  | Done B.ByteString
+  -- ^ The decoder has ended with leftover input.
+  | Error B.ByteString String
+  -- ^ The decoder has encountered an error with lefover input and an error
+  -- message.
+
+-- | Push an input chunk to the decoder
+pushChunk :: Decoder a -> B.ByteString -> Decoder a
+pushChunk decoder chunk = case decoder of
+  Consume k -> k chunk
+  Produce a decoder' -> Produce a $ decoder' `pushChunk` chunk
+  Done leftover -> Done $ leftover `B.append` chunk
+  Error leftover err -> Error (leftover `B.append` chunk) err
+
+-- | Decode a header and continue with the provided decoder
+withHeader
+  :: (Header -> B.ByteString -> Decoder r)
+  -- ^ Continuation
+  -> Decoder r
+withHeader f = go $ G.runGetIncremental getHeader
+  where
+    go decoder = case decoder of
+      G.Done leftover _ header -> f header leftover
+      G.Partial k -> Consume $ \chunk -> go $ k $ Just chunk
+      G.Fail leftover _ err -> Error leftover err
+
+-- | Decode a header
+decodeHeader :: Decoder Header
+decodeHeader = withHeader $ \header leftover -> Produce header $ Done leftover
+
+-- | Decode events
+decodeEvents :: Header -> Decoder Event
+decodeEvents header = go (0 :: Int) Nothing decoder0
+  where
+    decoder0 = mkEventDecoder header
+    go !remaining !blockCap decoder = case decoder of
+      G.Done leftover consumed r -> do
+        let !decoder' = decoder0 `G.pushChunk` leftover
+        case r of
+          Just event -> case evSpec event of
+            EventBlock {..} ->
+              go (fromIntegral block_size) (mkCap cap) decoder'
+            _ -> do
+              let
+                !remaining' = remaining - fromIntegral consumed
+                !blockCap' = if remaining' > 0 then blockCap else Nothing
+                !event' = event { evCap = blockCap }
+              Produce event' $ go remaining' blockCap' decoder'
+          Nothing -> go remaining blockCap decoder'
+      G.Partial k ->
+        Consume $ \chunk -> go remaining blockCap $ k $ Just chunk
+      G.Fail leftover _ err ->
+        Error leftover err
+
+-- | Decode a header and events
+decodeEventLog :: Decoder Event
+decodeEventLog = withHeader $ \header leftover ->
+  decodeEvents header `pushChunk` leftover
+
+-- | Read a header from a lazy bytestring and return the header and the
+-- leftover input for subsequent decoding.
+readHeader :: BL.ByteString -> Either String (Header, BL.ByteString)
+readHeader = go $ Left decodeHeader
+  where
+    go r bytes = case r of
+      Left decoder -> case decoder of
+        Produce header decoder' -> case decoder' of
+          Done leftover -> Right (header, BL.Chunk leftover bytes)
+          _ -> fail "readHeader: unexpected decoder"
+        Consume k -> case bytes of
+          BL.Empty -> fail "readHeader: not enough bytes"
```

Comment:
Does this mean that a caller has to supply a ByteString containing (at least) all of the header in one go? In other words, is reading the header non-incremental?